### PR TITLE
Consider cpu-device proximity

### LINKED ISSUE #763: support mapping policies from processes to devices/ports
Allow the user to specify which device/s to use during the run according to a certain policy - nearest device to the used CPU / first device / do a round-robin between the devices etc.

### PR DESCRIPTION
**Pull request by eladpe**
fixes #763

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: src/ucs/config/global_opts.c [hunk 1]
```diff
@@ -162,6 +163,9 @@ static ucs_config_field_t ucs_global_opts_table[] = {
    "Maximal size of profiling log. New records will replace old records.",
    ucs_offsetof(ucs_global_opts_t, profile_log_size), UCS_CONFIG_TYPE_MEMUNITS},
 #endif
+  {"PROXIMITY_PRIO", "y",
+   "Consider PCI proximity when looking for HCA.\n",
```
Comment (shamisp) [inline src/ucs/config/global_opts.c]: HCA is IB specific. Please use more generic term - "device"

File: src/ucs/config/global_opts.h [hunk 2]
```diff
@@ -90,6 +90,9 @@ typedef struct {
     /* Limit for profiling log size */
      size_t                   profile_log_size;
 
+    /* Give priority for close HCA */
```
Comment (shamisp) [inline src/ucs/config/global_opts.h]: device

**Comment by yosefe** [comment]
ok to test

File: src/ucs/config/global_opts.c (hunk 1 above)
Comment (yosefe) [inline src/ucs/config/global_opts.c]: this should be put in IB memory domain configuration, not if UCS (which is a generic infra library)

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -701,6 +703,19 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
         return UCS_ERR_IO_ERROR;
     }
 
+    if (ucs_global_opts.proximity_prio) {
+        ret = sched_getaffinity(0, sizeof(process_affinity), &process_affinity);
+        if (ret) {
+            ucs_warn("sched_getaffinity() failed: %m");
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: ucs_error

File: src/uct/ib/base/ib_iface.c [hunk 4]
```diff
@@ -701,6 +703,19 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
...
+            return UCS_ERR_INVALID_PARAM;
+        }
+        /* update latency for remote device */
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: pls add space line

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1174/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3113/ for details (Mellanox internal link).

File: src/ucs/config/global_opts.c (hunk 1 above)
Comment (shamisp) [inline src/ucs/config/global_opts.c]: Agreed

**Comment by eladpe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1183/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3125/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3126/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3127/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3128/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3129/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1184/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1185/ for details.

File: src/uct/ib/base/ib_md.c [hunk 5]
```diff
@@ -125,6 +125,10 @@ static ucs_config_field_t uct_ib_md_config_table[] = {
    "  <priority>  - (optional) device priority, integer.\n",
    ucs_offsetof(uct_ib_md_config_t, custom_devices), UCS_CONFIG_TYPE_STRING_ARRAY},
 
+  {"PROXIMITY_PRIO", "y",
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: pls rename to PREFER_NEAREST (incl. the names of internal variables)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3130/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1186/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1187/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3131/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3132/ for details (Mellanox internal link).

**Comment by eladpe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1193/ for details.

**Comment by eladpe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1194/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3140/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1198/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1199/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3145/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3146/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3147/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3150/ for details (Mellanox internal link).

**Pull request merged**